tools/chpl-language-server/test/end-marker-inlays.py

```diff
@@ -0,0 +1,204 @@
+async def test_end_markers_functions(client: LanguageClient):
+    """
+    Test that end markers show up properly for functions.
+    """
+    file = """
+            proc foo() {
+
+
+            }
+            proc bar() {
+              proc nested() {
+
+
+              }
+            } //
+            proc foobar () do ;
+            proc foobar2 () {
+
+            }
+            """
+
+    inlays = [
+        (pos((3, 1)), "proc foo()"),
+        (pos((8, 3)), "proc nested()"),
+    ]
+
+    with source_file(client, file) as doc:
+        await check_end_marker_inlay_hints(
+            client, doc, rng((0, 0), endpos(file)), inlays
+        )
+        await save_file(client, doc)
+        assert len(client.diagnostics[doc.uri]) == 0
+
+
+@pytest.mark.asyncio
+async def test_end_markers_named_decl(client: LanguageClient):
+    """
+    Test that end markers show up properly for named decls.
+    """
+    file = """
+            module M {
+              record R {
+
+
+              }
+              class C {
+
+
+              }
+              enum E {
+                a, b
+
+              }
+              union U {
+                var a: int;
+                var b: int;
+              }
+            }
+            """
+
+    inlays = [
+        (pos((4, 3)), "record R"),
+        (pos((8, 3)), "class C"),
+        (pos((12, 3)), "enum E"),
+        (pos((16, 3)), "union U"),
+        (pos((17, 1)), "module M"),
+    ]
+
+    with source_file(client, file) as doc:
+        await check_end_marker_inlay_hints(
+            client, doc, rng((0, 0), endpos(file)), inlays
+        )
+        await save_file(client, doc)
+        assert len(client.diagnostics[doc.uri]) == 0
+
+
+@pytest.mark.asyncio
+async def test_end_markers_select(client: LanguageClient):
+    """
+    Test that end markers show up properly for select/when.
+    """
+    file = """
+            var x = 10;
+            select x {
+              when 10 {
+
+
+              } when 20 {
+
+
+              }
+            }
+            """
+
+    inlays = [(pos((8, 3)), "when 20"), (pos((9, 1)), "select x")]
+
+    with source_file(client, file) as doc:
+        await check_end_marker_inlay_hints(
+            client, doc, rng((0, 0), endpos(file)), inlays
+        )
+        await save_file(client, doc)
+        assert len(client.diagnostics[doc.uri]) == 0
+
+
+@pytest.mark.asyncio
+async def test_end_markers_loops(client: LanguageClient):
+    """
+    Test that end markers show up properly for loops.
+    """
+    file = """
+            for i in 1..10 {
+
+
+            }
+            for i in 1..10 do ;
+            [(i, j) in zip(1..10, 1..10)] {
+              ;
+              ;
+            }
+            while false {
+
+
+            }
+            var a: int;
+            forall i in 1..10 with (ref a) {
+
+
+            }
+            """
+
+    inlays = [
+        (pos((3, 1)), "for i in 1..10"),
+        (pos((8, 1)), "(i, j) in zip(1..10, 1..10)"),
```

Comment:
Should this have the square brackets? My intuition is that the "[" serves as the "for" etc., and the "for" loop marker includes the keyword.

Reply:
The reason it doesn't is because of `with` clauses. They do not currently get rendered in end markers, so it felt weird to include the closing bracket and not the `with` clause on loops.